File: java/src/main/java/br/com/emendes/removeduplicatesfromsortedarray/RemoveDuplicatesFromSortedArray.java
package br.com.emendes.removeduplicatesfromsortedarray;

/**
 * Dado um array inteiro nums ordenado em ordem crescente,
 * remova as duplicatas do array de forma que cada elemento único apareça apenas uma vez.
 * A ordem relativa dos elementos deve ser mantida a mesma.<br><br>
 * <p>
 * Como é impossível alterar o comprimento do array em algumas linguagens,
 * você deve colocar o resultado na primeira parte do array nums.
 * Mais formalmente, se houver k elementos após a remoção das duplicatas,
 * os primeiros k elementos de nums devem conter o resultado final.
 * Não importa o que você deixa além dos primeiros k elementos.<br><br>
 * <p>
 * Retorne k após colocar o resultado final nos primeiros k slots de nums.<br><br>
 * <p>
 * Não aloque espaço extra para outro array.
 * Você deve fazer isso modificando a matriz de entrada no local com memória extra O(1).<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= nums.length <= 3 * 10^4<br>
 * -100 <= nums[i] <= 100<br>
 * nums é ordenado em ordem crescente.
 */
public class RemoveDuplicatesFromSortedArray {

  // Complexidade de tempo O(n)
  // Complexidade de espaço O(1)
  public int removeDuplicates(int[] nums) {
    int numsLength = nums.length;

    int currIndex = 1;
    int lastIndex = 0;

    while (currIndex < numsLength) {
      if (nums[lastIndex] != nums[currIndex]) {
        nums[++lastIndex] = nums[currIndex];
      }

      currIndex++;
    }

    return lastIndex + 1;
  }

}

File: javascript/problems/RemoveDuplicatesFromSortedArray.js
/**
 * Dado um array de inteiros nums ordenado em ordem não decrescente remova os elementos duplicados in-place
 * de modo que cada elemento único apareça uma única vez. A ordem relativa dos elementos deve ser mantida a mesma.
 * Então retorne o número de elementos exclusivos em nums.
 *
 * Considere o número de elementos únicos de nums como k. Para ser aceito, você precisa fazer o seguinte:
 *
 * - Altere o array nums de modo que os primeiros k elementos de nums contenham os elementos únicos na
 * ordem em que estavam presentes em nums inicialmente.
 * Os elementos restantes de nums não são importantes, assim como o tamanho de nums.
 *
 * - Retorne k.
 *
 * Restrições:
 *
 * 1 <= nums.length <= 3 * 10⁴
 *
 * -100 <= nums[i] <= 100
 *
 * nums está ordenado em ordem não descrescente.
 *
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function (nums) {
  let k = 0;
  for (let i = 1; i < nums.length; i++) {
    if (nums[i] != nums[k]) {
      nums[++k] = nums[i];
    }
  }

  return k + 1;
};
